network/router.go

```diff
@@ -0,0 +1,394 @@
+
+	host Host
+	// connections keeps track of all active connections with the translation
+	// use of an array because it happens that some connections are opened at
+	// the same time on both endpoints, and thus registered one after the other,
+	// erasing the first conn.
+	connections map[ServerIdentityID][]Conn
+	connsMut    sync.Mutex
+
+	// boolean flag indicating that the router is already clos{ing,ed}
+	isClosed  bool
+	closedMut sync.Mutex
+
+	// we wait that all handleConn routines are done
+	wg    sync.WaitGroup
+	wgMut sync.Mutex
+}
+
+// NewRouter returns a fresh Router giving its identity, and the host we want to
+// use.
+func NewRouter(own *ServerIdentity, h Host) *Router {
+	r := &Router{
+		id:          own,
+		connections: make(map[ServerIdentityID][]Conn),
+		host:        h,
+		Dispatcher:  NewBlockingDispatcher(),
+	}
+	r.address = h.Address()
+	r.host.Listening()
+	h.Listening()
+	return r
+}
+
+// Start will start the listening routine of the underlying Host. It is a
+// blocking call until the listening is done (by calling r.Stop()).
+func (r *Router) Start() {
+	// The function given to the listener  does the exchange of ServerIdentity
+	// and pass the connection along to the router.
+	err := r.host.Listen(func(c Conn) {
+		dst, err := r.exchangeServerIdentity(c)
+		if err != nil {
+			log.Error("ExchangeServerIdentity failed:", err)
+			if err := c.Close(); err != nil {
+				log.Error("Couldn't close secure connection:",
+					err)
+			}
+			return
+		}
+		// pass it along
+		r.launchHandleRoutine(dst, c)
+	})
+	if err != nil {
+		log.Error("Error listening:", err)
+	}
+}
+
+// Stop will stop the listening routine, and stop any routine of handling
+// connections. Calling r.Start(), then r.Stop() then r.Start() again leads to
+// an undefined behaviour. Callers should most of the time re-create a fresh
+// Router.
+func (r *Router) Stop() error {
+	var err error
+	if r.host.Listening() {
+		err = r.host.Stop()
+	}
+	err2 := r.stopHandling()
+	r.reset()
+	if err != nil {
+		return err
+	} else if err2 != nil {
+		return err2
+	}
+	return nil
+}
+
+// Send sends to an ServerIdentity without wrapping the msg into a SDAMessage
+func (r *Router) Send(e *ServerIdentity, msg Body) error {
+	if msg == nil {
+		return errors.New("Can't send nil-packet")
+	}
+
+	c := r.connection(e.ID)
+	if c == nil {
+		var err error
+		c, err = r.connect(e)
+		if err != nil {
+			return err
+		}
+	}
+
+	log.Lvlf4("%s sends to %s msg: %+v", r.address, e, msg)
+	var err error
+	err = c.Send(msg)
+	if err != nil {
+		log.Lvl2(r.address, "Couldn't send to", e, ":", err, "trying again")
+		c, err := r.connect(e)
+		if err != nil {
+			return err
+		}
+		err = c.Send(msg)
+		if err != nil {
+			return err
+		}
+	}
+	log.Lvl5("Message sent")
+	return nil
+}
+
+func (r *Router) connect(si *ServerIdentity) (Conn, error) {
+	c, err := r.host.Connect(si.Address)
+	if err != nil {
+		return nil, err
+	}
+	if err := r.negotiateOpen(si, c); err != nil {
```

Comment:
After-discussion-conclusion: let's keep it in here, but remove the sending of the identity from the listener perspective, since the connect-er knows already the listener identity since he is connecting to it.